### PR TITLE
Support HMAC signing on artifact uploads

### PR DESCRIPTION
**Pull request by gaspar09**
Introduce RemoteCacheOptions field to turbo.json
Adds initial test scaffolding for turbo json config parsing.
Implement HMAC Signature sent  on uploads to the remote cache via x-artifact-tag header.
Implement Signature Stream Validator on artifact downloads.
Download full response body and verify before untar when the signature/verification is enabled.

In followup: 
  - Restrict uploads to `teamId` specified in `remoteCacheOptions`

### CODE REVIEW

File: cli/internal/fs/package_json.go [hunk 1]
```diff
@@ -14,11 +14,10 @@ type TurboConfigJSON struct {
 	// Base Git branch
 	Base string `json:"baseBranch,omitempty"`
 	// Global root filesystem dependencies
-	GlobalDependencies []string `json:"globalDependencies,omitempty"`
-	TurboCacheOptions  string   `json:"cacheOptions,omitempty"`
-	Outputs            []string `json:"outputs,omitempty"`
-	// RemoteCacheUrl is the Remote Cache API URL
-	RemoteCacheUrl string `json:"remoteCacheUrl,omitempty"`
```
Comment (gaspar09) [inline cli/internal/fs/package_json.go:21]: @jaredpalmer , I notice this `RemoteCacheUrl` is not used anywhere. Is it okay to remove?

Comment (jaredpalmer) [reply cli/internal/fs/package_json.go:21]: Yes

File: cli/internal/fs/package_json.go [hunk 2]
```diff
@@ -14,11 +14,10 @@ type TurboConfigJSON struct {
...
+	GlobalDependencies []string           `json:"globalDependencies,omitempty"`
+	TurboCacheOptions  string             `json:"cacheOptions,omitempty"`
```
Comment (jaredpalmer) [inline cli/internal/fs/package_json.go:18]: Idk if these are used either?

**Comment by jaredpalmer** [comment]
annoying nit: we also need to update schema.d.ts w/ comments for our JSONSchema to play nice with VSCode/IDE automcomplete here: https://github.com/vercel/turborepo/blob/main/docs/schema.d.ts

There is already a PR by @Schniz to move it all to be Go, but Go isn't installed in build container on Vercel.

**Review by gsoltis — DISMISSED** [review]
It might be a good idea in a follow-on PR to include a test for a mismatch to ensure we are accounting for what happens in that scenario.

File: cli/internal/fs/testdata/turbo.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      "outputs": ["dist/**", ".next/**"]
+    },
+    "lint": {
+      "outputs": []
+    },
+    "dev": {
+      "cache": false
+    }
+  },
+  "remoteCacheOptions": {
```
Comment (jaredpalmer) [inline cli/internal/fs/testdata/turbo.json:14]: nit: aren't `Options` implied? Should it just be `remoteCache`

File: cli/internal/fs/testdata/turbo.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    "teamId": "team_id",
+    "signatureOptions": {
```
Comment (jaredpalmer) [inline cli/internal/fs/testdata/turbo.json:16]: can we just call this `signature`?

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

**Comment by jaredpalmer** [comment]
In another follow up, we should allow folks to set `cacheDirectory` and once #833 `remoteOnly`. Wondering if we want to refactor the payload to `cacheOptions`

**Comment by gaspar09** [comment]
> In another follow up, we should allow folks to set `cacheDirectory` and once #833 `remoteOnly`. Wondering if we want to refactor the payload to `cache`

noted: 
https://github.com/vercel/turborepo/issues/952

**Review by jaredpalmer — APPROVED** [review]

**Pull request merged**